activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb

```diff
@@ -311,9 +312,9 @@ def connection
       end
 
       def pin_connection!(lock_thread) # :nodoc:
-        raise "There is already a pinned connection" if @pinned_connection
```

Comment:
One thing I want to discuss is whether we want to check that the `lock_thread` value is consistent or not? I mean, how to treat the following scenarios:

```ruby
# 1

pin_connection!(true) #=> locked
pin_connection!(false) #=> still locked
unpin_conenction!
#=> locked 'till the end
...

# 2
pin_connection!(false) #=> unlocked
pin_connection!(true) #=> not it's locked
unpin_connection!
#=> still locked here! — do we want to be unlocked here?
...
```

So, right we keep the lock once it was set 'till the very final `#unpin_connection!`.

I can add a stack to keep locked values and restore in the `#unpin_connection!` accordingly, but.. do we need such complication?

Alternatively, we can treat non-locked pinned connections as _non-joinable_ with locked ones, i.e.:

```ruby
pin_connection!(false)
pin_connection!(false) # this is fine
pin_connection!(true) # !!! raises "Can not lock already pinned connection"

```

Or we can just ignore this scenario for now.

@byroot Thoughts?

Reply:
Yeha, that's the kind of weirdness I feared. That said, I think we can consider that code that don't ask for the lock, is likely OK with the lock existing. Meaning we can treat the lock as an upgrade:

  - Calling `pin_connection(true)` should add the lock if not already there
  - Calling `pin_connection(false)` should not remove the lock if there is one.
  - Only the final `unpin_connection` should remove the lock.

Reply:
Also I don't think we really want to create nested transactions. We should only call `begin_transaction` if `depth == 1`.